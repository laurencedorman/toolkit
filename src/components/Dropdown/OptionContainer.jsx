/**
 * Copyright (c) Colibri SAS - ManoMano
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import ToggleDown from '../ToggleDown';
import styles from './Dropdown.module.scss';

/**
 * @visibleName Dropdown
 */
const OptionContainer = ({
  on,
  toggle,
  size,
  options,
  right,
  className,
  onClick,
  active,
  sideLeft,
  sideRight,
}) => {
  const container = cn(styles.container, {
    [styles.show]: on,
    [styles.right]: right,
    [styles.sideLeft]: sideLeft,
    [styles.sideRight]: sideRight,
  });

  const itemOption = item =>
    cn(
      'li',
      className,
      { [styles.disabled]: item.disabled },
      { [styles.active]: (item.value || item.id) === active }
    );

  const list = cn('ul', styles.list);

  const space = 8;

  const setPosition = {
    top: size && !sideLeft && !sideRight && `${size.height + space}px`,
    right: size && sideLeft && !sideRight && `${size.width + space}px`,
    left: size && sideRight && `${size.width + space}px`,
  };

  return (
    <>
      {on && <div className={styles.closeTarget} onClick={toggle} />}
      <ToggleDown on={on} className={container} style={setPosition}>
        <ul className={list}>
          {options.map((item, index) => {
            const key = item.id ? item.id.toString() : index;
            return (
              <li
                key={key}
                className={itemOption(item)}
                data-id={item.id}
                data-value={item.value}
                disabled={item.disabled}
                onClick={!item.disabled ? onClick : null}
                onKeyPress={null}
              >
                <span
                  onClick={!item.disabled ? toggle : null}
                  onKeyPress={null}
                  key={key}
                >
                  {item.value}
                </span>
              </li>
            );
          })}
        </ul>
      </ToggleDown>
    </>
  );
};

OptionContainer.propTypes = {
  on: PropTypes.bool.isRequired,
  options: PropTypes.arrayOf(PropTypes.shape()).isRequired,
  toggle: PropTypes.func,
  right: PropTypes.bool,
  onClick: PropTypes.func,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  active: PropTypes.string,
  sideLeft: PropTypes.bool,
  sideRight: PropTypes.bool,
};

OptionContainer.defaultProps = {
  toggle: null,
  onClick: null,
  right: false,
  className: '',
  disabled: false,
  active: '',
  sideLeft: false,
  sideRight: false,
};

export default OptionContainer;
