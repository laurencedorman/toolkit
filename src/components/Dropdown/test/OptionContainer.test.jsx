/**
 * Copyright (c) Colibri SAS - ManoMano
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { shallow } from 'enzyme';
import OptionContainer from '../OptionContainer';

describe('OptionContainer', () => {
  let wrapper;

  const props = {
    options: [
      {
        label: 'test',
        value: 'test',
      },
    ],
  };

  beforeEach(() => {
    wrapper = shallow(<OptionContainer {...props} on={false} />);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render container if props on is true', () => {
    // When
    wrapper = shallow(<OptionContainer {...props} on />);
    // then
    expect(wrapper.find('ul')).toHaveLength(1);
  });

  it('should render overlay if props on is true', () => {
    // When
    wrapper = shallow(<OptionContainer {...props} on />);
    // then
    expect(wrapper.find('.closeTarget')).toHaveLength(1);
  });
});
