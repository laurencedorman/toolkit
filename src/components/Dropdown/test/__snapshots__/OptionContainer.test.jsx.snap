// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionContainer should match snapshot 1`] = `
<Fragment>
  <ToggleDown
    className="container"
    on={false}
    style={
      Object {
        "left": undefined,
        "right": undefined,
        "top": undefined,
      }
    }
  >
    <ul
      className="ul list"
    >
      <li
        className="li"
        data-value="test"
        key="0"
        onClick={null}
        onKeyPress={null}
      >
        <span
          key="0"
          onClick={null}
          onKeyPress={null}
        >
          test
        </span>
      </li>
    </ul>
  </ToggleDown>
</Fragment>
`;
